File: csharp/QbGifBuilder/QbGifBuilder.cs
namespace QbGif {
  using System;
  using System.IO;
  using System.Collections.Generic;
  using System.Drawing;
  using System.Drawing.Imaging;

  public enum QbGuidBuilderDelayPolicy {
    UNKNOWN, IMAGE_ONLY, GLOBAL, PER_FRAME
  }

  public class QbGifBuilder {

    private List<QbGifFrame> frameList;
    private QbGuidBuilderDelayPolicy delayPolicy = QbGuidBuilderDelayPolicy.IMAGE_ONLY;

    private int defaultDelayInMs = 100;

    private int nbRepetitions = 65535;

    // CTOR
    public QbGifBuilder() {
      frameList = new List<QbGifFrame>();
    }

    public void SetDefaultDelayInMs (int delayInMs) {
      this.defaultDelayInMs = delayInMs;
    }
    public void SetDelayPolicy (QbGuidBuilderDelayPolicy policy) {
      this.delayPolicy = policy;
    }

    public void SetNbRepetitions (int nbRepetitions) {
      if (nbRepetitions < 0) {
        nbRepetitions = 0;
      }
      if (nbRepetitions > 65535) {
        nbRepetitions = 0;
      }
      this.nbRepetitions = nbRepetitions;
    }

    public void AddFrame(Bitmap bitmap) {
      this.AddFrame( bitmap, this.defaultDelayInMs);
    }

    public void AddFrame(Bitmap bitmap, int delayInMs) {
      using (MemoryStream memStream = new MemoryStream()) {
        bitmap.Save(memStream, ImageFormat.Gif);
        memStream.Seek(0, SeekOrigin.Begin);
        List<QbGifFrame> frameList = QbGifFrame.CreateGifFrames(memStream);
        foreach (QbGifFrame frame in frameList) {
          frame.SetDelayInMs(delayInMs);
          this.frameList.Add(frame);
        }
      }
    }

    public void AddFrames(Stream stream) {
      AddFrames(stream, this.defaultDelayInMs);
    }

    public void AddFrames(Stream stream, int delayInMs) {
      List<QbGifFrame> frameList = QbGifFrame.CreateGifFrames(stream);
      // Apply delay policy
      if (frameList.Count > 0) {
        
        foreach (QbGifFrame frame in frameList) {
          if (frame.GetDelayInMs() == -1) {
            frame.SetDelayInMs( delayInMs );              
          }
        }

        switch ( this.delayPolicy) {

          case QbGuidBuilderDelayPolicy.PER_FRAME:
            foreach (QbGifFrame frame in frameList) {
              frame.SetDelayInMs(delayInMs);
            }
            break;              

          case QbGuidBuilderDelayPolicy.GLOBAL:
            // Calculate total duration
            int totalDuration = 0;
            foreach (QbGifFrame frame in frameList) {
              totalDuration += frame.GetDelayInMs();
            }
            // Recalculate all durations
            foreach (QbGifFrame frame in frameList) {
              int delay = delayInMs * frame.GetDelayInMs() / totalDuration;
              frame.SetDelayInMs(delay);
            }
            break;

          case QbGuidBuilderDelayPolicy.UNKNOWN:
          case QbGuidBuilderDelayPolicy.IMAGE_ONLY:
          default:
            // do nothing
            break;
      }

      // append new frame list to the QbGifBuilder
      foreach (QbGifFrame frame in frameList) {
          this.frameList.Add(frame);
        }
      }
    }

    public void Export(Stream outputStream) {
      
      if (this.frameList.Count == 0 ) {
        throw (new Exception("Cannot generate GIF. List is empty."));
      }

      // calculate maxWidth and minWidth
      int maxWidth = 0;
      int maxHeight = 0;
      foreach (QbGifFrame frame in this.frameList) {
        if (frame.GetGlobalWidth() > maxWidth) {
          maxWidth = frame.GetGlobalWidth();
        }
        if (frame.GetGlobalHeight() > maxHeight) {
          maxHeight = frame.GetGlobalHeight();
        }
      }
      QbGifFrame.WriteGlobalData(outputStream, maxWidth, maxHeight);
      QbGifFrame.WriteNetscape20Extension(outputStream, this.nbRepetitions);
      // Insert frames
      foreach (QbGifFrame frame in this.frameList) {
        frame.WriteLocalData(outputStream);
      }
      // terminator
      outputStream.WriteByte(0x3b);
    }

  }
}

File: cplusplus/QbGifBuilder/src/QbGifBuilder.cpp
#include "QbGifBuilder.h"

#include "QbGifFrame.h"
#include "QbStreamTools.h"

QbGifBuilder::QbGifBuilder() {
  // Default values
  this->defaultDelayInMs = 100;
  this->delayPolicy = IMAGE_ONLY;
  this->nbRepetitions = 65535;
}

QbGifBuilder::~QbGifBuilder() {
  for (auto frame : this->frameList) {
    delete(frame); 
  }
}

void QbGifBuilder::setDefaultDelayInMs (int delayInMs) {
  this->defaultDelayInMs = delayInMs;
}

void QbGifBuilder::setDelayPolicy (DelayPolicy policy) {
  this->delayPolicy = policy;
}

void QbGifBuilder::setNbRepetitions (int nbRepetitions) {
  if (nbRepetitions < 0) {
    nbRepetitions = 0;
  }
  if (nbRepetitions > 65535) {
    nbRepetitions = 0;
  }
  this->nbRepetitions = nbRepetitions;
}

void QbGifBuilder::addFrame(const QbGifBitmap *bitmap) {
  this->addFrame(bitmap, this->defaultDelayInMs);
}

void QbGifBuilder::addFrame(const QbGifBitmap *bitmap, int delayInMs) {
  QbGifFrame * frame = new QbGifFrame(bitmap);
  frame->setDelayInMs(delayInMs);
  this->frameList.push_back(frame);
}

void QbGifBuilder::addFrames(istream &stream) {
  this->addFrames(stream, this->defaultDelayInMs);
}

void QbGifBuilder::addFrames(istream &stream, int delayInMs) {
  vector<QbGifFrame *> frameList = QbGifFrame::createGifFrames(stream);
  // Apply delay policy
  if (frameList.size() > 0) {
  for (QbGifFrame * frame : frameList) {
      if (frame->getDelayInMs() == -1) {
        frame->setDelayInMs( delayInMs );              
      }      
    }

    switch ( this->delayPolicy) {

      case PER_FRAME: {
        for (QbGifFrame * frame : frameList) {
          frame->setDelayInMs(delayInMs);
        }
        break;
      }     

      case GLOBAL: {
        // Calculate total duration
        int totalDuration = 0;
        for (QbGifFrame * frame : frameList) {
            totalDuration += frame->getDelayInMs();
        }
        // Recalculate all durations
        for (auto frame : frameList) {
          int delay = delayInMs * frame->getDelayInMs() / totalDuration;
          frame->setDelayInMs(delay);
        }
        break;
      }

      case UNKNOWN:
      case IMAGE_ONLY:      
      default: {
        // do nothing
        break;
      }

    }
    // append new frame list to the QbGifBuilder
    for (QbGifFrame * frame : frameList) {
      this->frameList.push_back(frame);
    }
  }
}

void QbGifBuilder::exportGif(ostream &outputStream) { 
  if (this->frameList.size() == 0 ) {
    cout << "Cannot generate GIF. List is empty.\n";
    return;
  }

  // calculate maxWidth and minWidth
  int maxWidth = 0;
  int maxHeight = 0;
  for (auto frame : this->frameList) {
    if (frame->getGlobalWidth() > maxWidth) {
      maxWidth = frame->getGlobalWidth();
    }
    if (frame->getGlobalHeight() > maxHeight) {
      maxHeight = frame->getGlobalHeight();
    }
  }
  QbGifFrame::writeGlobalData(outputStream, maxWidth, maxHeight);
  QbGifFrame::writeNetscape20Extension(outputStream, this->nbRepetitions);
  // Insert frames
  for (auto frame : this->frameList) {
    frame->writeLocalData(outputStream);
  }
  // terminator
  QbStreamTools::writeByte(outputStream, 0x3b);
}
